util/src/main/java/io/grpc/util/HealthProducerUtil.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2023 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.LoadBalancer.HAS_HEALTH_PRODUCER_LISTENER_KEY;
+import static io.grpc.LoadBalancer.HEALTH_CONSUMER_LISTENER_ARG_KEY;
+
+import io.grpc.Attributes;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.Internal;
+import io.grpc.LoadBalancer;
+
+/**
+ * Utility function used by health producer systems to build health notification chain.
+ * The leaf health consumer is pick first. Each health producer using this helper.
+ * The most root health producer in the chain will fan out the subchannel state change to both
+ * state listener and health listener.
+ */
+@Internal
+public class HealthProducerUtil {
```

Comment:
oh i see what you meant.  These two classes are tightly related so I would put them together instead of two. I am not too worries about this.
But I will put it in api package and also have two attributes moved to this class.

Reply:
oh i might still put the HealthCheckUtil in util because it depends on ForwardingSubchannel and ForwardingLoadBalancerHelper that are in util. `:grpc-api` does not depends on `:grpc-util`

Reply:
I guess this is a style thing and about consistency. We have a lot of classes that are closely related that have defined independently in their own files. Should we start wrapping them "util" classes as well?

Also the "util" name usually refers to a class with a collection of static methods that the designer couldn't find proper homes for.

If you really don't want two class files you could also make one HealthProducerHelper.java that, as its internal concern, defines the subchannel as an inner class.